src/Optimization/hiopKKTLinSys.cpp

```diff
@@ -1473,6 +1614,318 @@ bool hiopKKTLinSysFull::computeDirections(const hiopResidual* resid,
+  yrd_->startingAtCopyFromStartingAt(0,   x, nx);
+  yryc_->startingAtCopyFromStartingAt(0,  x, nx+nineq);
+  yryd_->startingAtCopyFromStartingAt(0,  x, nx+nineq+neq);
+  yrsxl_->startingAtCopyFromStartingAt(0, x, nx+nineq+neq+nineq);
+  yrsxu_->startingAtCopyFromStartingAt(0, x, nx+nineq+neq+nineq+nx);
+  yrsdl_->startingAtCopyFromStartingAt(0, x, nx+nineq+neq+nineq+nx+nx);
+  yrsdu_->startingAtCopyFromStartingAt(0, x, nx+nineq+neq+nineq+nx+nx+nineq);
+  yrzl_->startingAtCopyFromStartingAt(0,  x, nx+nineq+neq+nineq+nx+nx+nineq+nineq);
+  yrzu_->startingAtCopyFromStartingAt(0,  x, nx+nineq+neq+nineq+nx+nx+nineq+nineq+nx);
+  yrvl_->startingAtCopyFromStartingAt(0,  x, nx+nineq+neq+nineq+nx+nx+nineq+nineq+nx+nx);
+  yrvu_->startingAtCopyFromStartingAt(0,  x, nx+nineq+neq+nineq+nx+nx+nineq+nineq+nx+nx+nineq);
+  return true;
+}
+
+bool hiopCompressPrecondOpr::combine_res_to_build_y(hiopVector& y)
+{
+  size_type nx = dx_->get_size();
+  size_type neq = dyc_->get_size();
+  size_type nineq = dyd_->get_size();
+
+  yrx_->copyToStarting(   y, 0);
+  yrd_->copyToStarting(   y, nx);
+  yryc_->copyToStarting(  y, nx+nineq);
+  yryd_->copyToStarting(  y, nx+nineq+neq);
+  yrsxl_->copyToStarting( y, nx+nineq+neq+nineq);
+  yrsxu_->copyToStarting( y, nx+nineq+neq+nineq+nx);
+  yrsdl_->copyToStarting( y, nx+nineq+neq+nineq+nx+nx);
+  yrsdu_->copyToStarting( y, nx+nineq+neq+nineq+nx+nx+nineq);
+  yrzl_->copyToStarting(  y, nx+nineq+neq+nineq+nx+nx+nineq+nineq);
+  yrzu_->copyToStarting(  y, nx+nineq+neq+nineq+nx+nx+nineq+nineq+nx);
+  yrvl_->copyToStarting(  y, nx+nineq+neq+nineq+nx+nx+nineq+nineq+nx+nx);
+  yrvu_->copyToStarting(  y, nx+nineq+neq+nineq+nx+nx+nineq+nineq+nx+nx+nineq);
+  return true;
+}
+
+bool hiopCompressPrecondOpr::combine_dir_to_build_y(hiopVector& y)
+{
+  size_type nx = dx_->get_size();
+  size_type neq = dyc_->get_size();
+  size_type nineq = dyd_->get_size();
+
+  dx_->copyToStarting(   y, 0);
+  dd_->copyToStarting(   y, nx);
+  dyc_->copyToStarting(  y, nx+nineq);
+  dyd_->copyToStarting(  y, nx+nineq+neq);
+  dsxl_->copyToStarting( y, nx+nineq+neq+nineq);
+  dsxu_->copyToStarting( y, nx+nineq+neq+nineq+nx);
+  dsdl_->copyToStarting( y, nx+nineq+neq+nineq+nx+nx);
+  dsdu_->copyToStarting( y, nx+nineq+neq+nineq+nx+nx+nineq);
+  dzl_->copyToStarting(  y, nx+nineq+neq+nineq+nx+nx+nineq+nineq);
+  dzu_->copyToStarting(  y, nx+nineq+neq+nineq+nx+nx+nineq+nineq+nx);
+  dvl_->copyToStarting(  y, nx+nineq+neq+nineq+nx+nx+nineq+nineq+nx+nx);
+  dvu_->copyToStarting(  y, nx+nineq+neq+nineq+nx+nx+nineq+nineq+nx+nx+nineq);
+  return true;
+}
+
+bool hiopCompressPrecondOpr::times_vec(hiopVector& y, const hiopVector& x)
+{
+  bool bret;
+//  bret = split_x_to_build_it(x);
+  bret = split_x_to_build_res(x);
+
+  bret = kkt_->computeDirections(resid_, dir_); 
+
+//  bret = combine_res_to_build_y(y);
+  bret = combine_dir_to_build_y(y);
+  return true;
+}
+
+bool hiopCompressPrecondOpr::trans_times_vec(hiopVector& y, const hiopVector& x)
```

Comment:
It is not called by BiCGStab. BiCGStab only needs the solution of the preconditioned system, i.e., KKT/rhs (hiopCompressPrecondOpr::times_vec). I will push a change where I call `times_vec` since compressed KKT is symmetric.